pkg/process/monitor_linux.go

```diff
@@ -0,0 +1,259 @@
+
+type deathEvent struct {
+	daemonID string
+	path     string
+}
+
+func newMonitor() (_ *livenessMonitor, err error) {
+	var epollFd int
+	epollFd, err = unix.EpollCreate1(unix.EPOLL_CLOEXEC)
+	if err != nil {
+		return nil, errors.Wrap(err, "create daemons monitor")
+	}
+
+	defer func() {
+		if err != nil {
+			unix.Close(epollFd)
+		}
+	}()
+
+	m := &livenessMonitor{
+		epollFd:     epollFd,
+		subscribers: make(map[string]*target),
+		set:         make(map[uintptr]*target),
+	}
+
+	return m, nil
+}
+
+func (m *livenessMonitor) Subscribe(id string, path string, notifier chan<- deathEvent) (err error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if s, ok := m.subscribers[id]; ok && s.id == path {
+		log.L.Warnf("Daemon %s is already subscribed!", id)
+		return errdefs.ErrAlreadyExists
+	}
+
+	var (
+		c       net.Conn
+		rawConn syscall.RawConn
+	)
+
+	err = retry.Do(func() (err error) {
+		// Don't forget close me!
+		if c, err = net.Dial("unix", path); err != nil {
+			log.L.Errorf("Fails to connect to %s, %v", path, err)
+			return
+		}
+
+		return nil
+	},
+		retry.Attempts(4),
+		retry.LastErrorOnly(true),
+		retry.Delay(100*time.Millisecond))
+
+	if err != nil {
+		return err
+	}
+
+	uc, ok := c.(*net.UnixConn)
+	if !ok {
+		return errors.Errorf("a unix socket connection is required")
+	}
+
+	if rawConn, err = uc.SyscallConn(); err != nil {
+		return
+	}
+
+	err = rawConn.Control(func(fd FD) {
+		err = unix.SetNonblock(int(fd), true)
+		if err != nil {
+			log.L.Errorf("Failed to set file. daemon id %s path %s. %v", id, path, err)
+			return
+		}
+
+		event := unix.EpollEvent{
+			Fd:     int32(fd),
+			Events: unix.EPOLLHUP | unix.EPOLLERR | unix.EPOLLET,
+		}
+
+		err = unix.EpollCtl(m.epollFd, unix.EPOLL_CTL_ADD, int(fd), &event)
+		if err != nil {
+			log.L.Errorf("Failed to control epoll. daemon id %s path %s. %v", id, path, err)
+			return
+		}
+		target := &target{uc: uc, id: id, path: path}
+
+		// Only add subscribed target when everything is OK.
+		m.set[fd] = target
+		m.subscribers[id] = target
+		target.notifier = notifier
+	})
+
+	log.L.Infof("Subscribe daemon %s liveness event, path=%s.", id, path)
+
+	return
+}
+
+func (m *livenessMonitor) Unsubscribe(id string) (err error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	return m.unsubscribe(id)
+}
+
+func (m *livenessMonitor) unsubscribe(id string) (err error) {
+	target, ok := m.subscribers[id]
+	if !ok {
+		return errdefs.ErrNotFound
+	}
+
+	delete(m.subscribers, id)
+
+	var rawConn syscall.RawConn
+	if rawConn, err = target.uc.SyscallConn(); err != nil {
+		log.L.Errorf("Fail to access underlying FD, id=%s", id)
+		return
+	}
+
+	// No longer wait for event, delete it from interest list.
+	if err = rawConn.Control(func(fd uintptr) {
+		unix.EpollCtl(m.epollFd, unix.EPOLL_CTL_DEL, int(fd), &unix.EpollEvent{})
+		delete(m.set, fd)
+	}); err != nil {
+		return errors.Wrapf(err, "remove target FD in the interested list, id=%s", id)
+	}
+
+	if err = target.uc.Close(); err != nil {
```

Comment:
Maybe give a warning for the error.

Reply:
Sure, though the error is thrown up already

Reply:
fixed